dhis-2/dhis-support/dhis-support-db-migration/src/main/java/org/hisp/dhis/db/migration/callback/BeforeMigration_WrapMetadataVersion.java

```diff
@@ -0,0 +1,311 @@
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement;
+import org.flywaydb.core.api.callback.Callback;
+import org.flywaydb.core.api.callback.Context;
+import org.flywaydb.core.api.callback.Event;
+import org.hisp.dhis.common.DxfNamespaces;
+import org.hisp.dhis.metadata.version.MetadataVersion;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * @author David Katuscak
+ */
+public class BeforeMigration_WrapMetadataVersion implements Callback
+{
+    private static final Logger log = LoggerFactory.getLogger( BeforeMigration_WrapMetadataVersion.class );
+
+    private static final String WRAPPED_METADATA_PREFIX = "{\"metadata\":";
+
+    private static final String VALUE_COLUMN_EXISTS_SQL = "SELECT count(0) FROM information_schema.columns " +
+        "WHERE table_schema = 'public' AND table_name = 'keyjsonvalue' AND column_name = 'value';";
+
+    private static final String GET_ALL_METADATA_VERSIONS_SQL = "SELECT uid, name, hashcode FROM metadataversion";
+    private static final String UPDATE_METADATA_VERSION_SQL = "UPDATE metadataversion SET hashcode = ? WHERE uid = ?";
+    private static final String GET_KEYJSONVALUE_SQL = "SELECT value, uid FROM keyjsonvalue WHERE namespace='METADATASTORE' AND namespacekey = ?";
+    private static final String UPDATE_KEYJSONVALUE_SQL = "UPDATE keyjsonvalue SET value = ? WHERE uid = ?";
+
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    static
+    {
+        mapper.enableDefaultTyping();
+        mapper.setSerializationInclusion( JsonInclude.Include.NON_NULL );
+    }
+
+    @Override
+    public boolean supports( final Event event, final Context context )
+    {
+        return event == Event.BEFORE_MIGRATE;
+    }
+
+    @Override
+    public boolean canHandleInTransaction( final Event event, final Context context )
+    {
+        return true;
+    }
+
+    @Override
+    public void handle( final Event event, final Context context )
+    {
+        try
+        {
+            if ( necessaryToRunMigration( context ) )
```

Comment:
Maybe change the idempotency check based on the flyway_schema_history table rather than the information schema table